packages/commonwealth/client/scripts/state/api/comments/deleteComment.ts

```diff
@@ -0,0 +1,95 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import axios from 'axios';
+import app from 'state';
+import { ApiEndpoints } from 'state/api/config';
+import { updateThreadInAllCaches } from '../threads/helpers/cache';
+import useFetchCommentsQuery from './fetchComments';
+
+interface DeleteCommentProps {
+  address: string;
+  chainId: string
+  canvasHash: string
+  commentId: number;
+  existingNumberOfComments: number;
+}
+
+const deleteComment = async ({
+  address,
+  chainId,
+  commentId,
+  canvasHash,
+}: DeleteCommentProps) => {
+  const {
+    session = null,
+    action = null,
+    hash = null,
+  } = await app.sessions.signDeleteComment({
+    comment_id: canvasHash,
+  })
+
+  await axios.delete(`${app.serverUrl()}/comments/${commentId}`, {
+    data: {
+      jwt: app.user.jwt,
+      address: address,
+      chain: chainId,
+      author_chain: chainId
+    },
+  })
+
+  // Important: we render comments in a tree, if this deleted comment was
+  // the root comment of a tree, then we want to preserve the comment tree,
+  // but in place of this deleted comment we will show the "[deleted]" msg.
+  return {
+    softDeleted: {
+      id: commentId,
+      deleted: true,
+      text: '[deleted]',
+      plaintext: '[deleted]',
+      versionHistory: [],
+      canvas_action: action,
+      canvas_session: session,
+      canvas_hash: hash,
+    }
+  }
+};
+
+interface UseDeleteCommentMutationProps {
+  chainId: string
+  threadId: number;
+  existingNumberOfComments: number;
+}
+
+const useDeleteCommentMutation = ({ chainId, threadId, existingNumberOfComments }: UseDeleteCommentMutationProps) => {
+  const queryClient = useQueryClient();
+  const { data: comments } = useFetchCommentsQuery({
+    chainId,
+    threadId,
+  })
+
+  return useMutation({
+    mutationFn: deleteComment,
+    onSuccess: async (response) => {
+      // find the existing comment index
+      const foundCommentIndex = comments.findIndex(x => x.id === response.softDeleted.id)
+
+      if (foundCommentIndex > -1) {
+        const softDeletedComment = Object.assign({ ...comments[foundCommentIndex] }, { ...response.softDeleted })
+
+        // update fetch comments query state
+        const key = [ApiEndpoints.FETCH_COMMENTS, chainId, threadId]
+        queryClient.cancelQueries({ queryKey: key });
+        queryClient.setQueryData([...key],
+          () => {
+            const updatedComments = [...(comments || [])]
+            updatedComments[foundCommentIndex] = { ...softDeletedComment }
+            return [...updatedComments]
```

Comment:
I think we had some discussion recently about redundant destructuring?

Reply:
If we remove the spread, we will be mutating the array, I already suggested before to not mutate as it would break the cache most of the time, since mutation will update in place instead of proper updation.

And I don't think we need to add a utility here, because 
A- this logic is not used anywhere else
B- over abstraction (please refer to old code where had a ton of unused code - which is already an example of over-abstraction)

I am happy to add a utility if you still suggest to add one, again just put my thoughts out here for better colab and heading to the best solution.

cc: @mzparacha @jnaviask

Reply:
I'm personally okay with this pattern for the time being -- we should think about whether there's a simpler solution, maybe just a small "deep copy remove" utility or something like that.